### PR TITLE
[DebugInfo][RemoveDIs] Add documentation for updating code to handle debug records

### PR DESCRIPTION
**Pull request by SLTozer**
Although the patch that enables debug records by default has been temporarily reverted, it will (eventually) be reverted and everyone's code will be subjected to the new debug info format. Although this is broadly a good thing, it is important that the documentation has enough information to guide users through the update; this patch adds what should hopefully be enough detail for most users to either find the answers, or find out how to find those answers.

### CODE REVIEW

**Comment by jryans** [comment]
What's the status of https://github.com/llvm/llvm-project/pull/91725? That seems to update some of the same text (at least the summary paragraph at the top). Maybe it was forgotten? Perhaps best to merge that first and then rebase this to avoid conflicts.

**Comment by SLTozer** [comment]
> What's the status of #91725? That seems to update some of the same text (at least the summary paragraph at the top). Maybe it was forgotten? Perhaps best to merge that first and then rebase this to avoid conflicts.

That patch has been deliberately left hanging until the [corresponding textual IR update](https://github.com/llvm/llvm-project/pull/91724) has landed/is about to land, just in case any change to the approach becomes necessary before then. It's more likely that this patch will land first, because _this_ patch exists to assist with https://github.com/llvm/llvm-project/pull/89799, which is currently reverted and waiting for a reduced reproducer, and the IR patch is blocked on that patch. So in short, it's just a dependency tree situation; this change should really have been made before landing the aforementioned PR in the first place.

File: llvm/docs/RemoveDIsDebugInfo.md [hunk 1]
```diff
@@ -106,21 +106,204 @@ Not shown are the links from DbgRecord to other parts of the `Value`/`Metadata`
 
 The various kinds of debug intrinsic (value, declare, assign, label) are all stored in `DbgRecord` subclasses, with a "RecordKind" field distinguishing `DbgLabelRecord`s from `DbgVariableRecord`s, and a `LocationType` field in the `DbgVariableRecord` class further disambiguating the various debug variable intrinsics it can represent.
 
-## Finding debug info records
+# How to update existing code
+
+Any existing code that interacts with debug intrinsics in some way will need to be updated to interact with debug records in the same way. A few quick rules to keep in mind when updating code:
+
+- Debug records will not be seen when iterating over instructions; to find the debug records that appear immediately before an instruction, you'll need to iterate over `Instruction::getDbgRecordRange()`.
+- Debug records have interfaces that are identical to those of debug intrinsics, meaning that any code that operates on debug intrinsics can be trivially applied to debug records as well. The exceptions for this are Instruction or CallInst methods that don't logically apply to debug records, and `isa/cast/dyn_cast` methods, are replaced by methods on the `DbgRecord` class itself.
```
Comment (jryans) [inline llvm/docs/RemoveDIsDebugInfo.md:114]:
```suggestion
- Debug records have interfaces that are identical to those of debug intrinsics, meaning that any code that operates on debug intrinsics can be trivially applied to debug records as well. The exceptions for this are `Instruction` or `CallInst` methods that don't logically apply to debug records, and `isa`/`cast`/`dyn_cast` methods, are replaced by methods on the `DbgRecord` class itself.
```

File: llvm/docs/RemoveDIsDebugInfo.md [hunk 2]
```diff
@@ -106,21 +106,204 @@ Not shown are the links from DbgRecord to other parts of the `Value`/`Metadata`
...
+  }
+  ```
+
+Below is a rough guide on how existing code that currently supports debug intrinsics can be updated to support debug records.
+
+## Creating debug records
+
+Debug records will automatically be created by the `DIBuilder` class when the new format is enabled. As with instructions, it is also possible to call `DbgRecord::clone` to create an unattached copy of an existing record.
+
+## Skipping debug records, ignoring debug-uses of Values, stably counting instructions...
```
Comment (jryans) [inline llvm/docs/RemoveDIsDebugInfo.md:134]:
```suggestion
## Skipping debug records, ignoring debug-uses of `Value`s, stably counting instructions, etc.
```

File: llvm/docs/RemoveDIsDebugInfo.md [hunk 3]
```diff
@@ -106,21 +106,204 @@ Not shown are the links from DbgRecord to other parts of the `Value`/`Metadata`
...
+      recordDeclareAddress(DVR.getAddress());
+  }
+  // ...
+}
+```
+
+## Processing individual debug records
 
-Erase `DbgRecord`s by calling `eraseFromParent` or `deleteInstr` if it's already been removed.
+In most cases, any code that operates on debug intrinsics can be extracted to a template function or auto lambda (if it is not already in one) that can be applied to both debug intrinsics and debug records - though keep in mind the main exception that isa/cast/dyn_cast do not apply to DbgVariableRecord types.
```
Comment (jryans) [inline llvm/docs/RemoveDIsDebugInfo.md:222]:
```suggestion
In most cases, any code that operates on debug intrinsics can be extracted to a template function or auto lambda (if it is not already in one) that can be applied to both debug intrinsics and debug records - though keep in mind the main exception that `isa`/`cast`/`dyn_cast` do not apply to `DbgVariableRecord` types.
```

**Review by jryans — APPROVED** [review]
Looks good overall, thanks for working on this! 😄

**Comment by dwblaikie** [comment]
> That patch has been deliberately left hanging until the https://github.com/llvm/llvm-project/pull/91724 has landed/is about to land, just in case any change to the approach becomes necessary before then.

Given that that patch only changes the default behavior - if someone wanted to get ahead of that default switch they'd want the documentation now rather than later. And if things change, we can update the documentation - might be worth submitting this now so it's not sort of floating out here all branch-like. But I don't feel super strongly about it.

**Comment by SLTozer** [comment]
> Given that that patch only changes the default behavior - if someone wanted to get ahead of that default switch they'd want the documentation now rather than later.

That's a reasonable argument - I'd just assumed that nobody was going to be proactively applying these changes, but it's easier to do so if the information is available sooner I suppose!

**Pull request merged**